src/Utilities/ContainerHelpers.hpp

```diff
@@ -3,43 +3,138 @@
+ * function `at`, which can be used to specify a custom indexing function. For
+ * instance, for a type which is a `std::array` of a `std::vector`, the indexing
+ * function could be the below callable struct:
+ * ```
+ * struct ArrayOfVectorIndexFunctor {
+ *   template <size_t ArraySize, typename T>
+ *   T& operator()(std::array<std::vector<T>, ArraySize> array_of_vector,
+ *              size_t index) noexcept {
+ *     return array_of_vector[index % ArraySize].at(index / ArraySize);
+ *   }
+ * };
+ * ```
+ * which would index the data structure in a manner in which the array index
+ * varies fastest. The indexing function must take as arguments the applicable
+ * container and a `size_t` index, in that order. This follows the convention of
+ * `gsl::at`.
+ * \note `std::complex` are regarded as non-indexable (despite a predictable
+ * memory layout), so this function acts as the identity on `std::complex` of
+ * fundamental types)
  */
-template <class T>
-SPECTRE_ALWAYS_INLINE auto get_element(T& t, const size_t i) noexcept
-    -> decltype(t[i]) {
-  return t[i];
+template <typename T, typename SubscriptFunction = At>
+SPECTRE_ALWAYS_INLINE decltype(auto) get_element(
+    T& t, size_t i, SubscriptFunction at = At{}) noexcept {
+  return ContainerHelpersDetail::GetElementImpl<T>{}(t, i, at);
 }
 
-/// \cond
-template <class T, Requires<std::is_fundamental<T>::value> = nullptr>
-SPECTRE_ALWAYS_INLINE T& get_element(T& t, size_t /*i*/) noexcept {
-  return t;
-}
-/// \endcond
-
 /*!
  * \ingroup UtilitiesGroup
  * \brief Retrieve the size of `t` if `t.size()` is a valid expression,
- * otherwise if `T` is fundamental returns 1
+ * otherwise if `T` is fundamental or a `std::complex` of a fundamental type,
+ * returns 1.
+ *
+ * \details This function also optionally takes the user-defined `size`
+ * function, which can be used to specify a custom size function. For
+ * instance, for a type which is a `std::array` of a `std::array`, the size
+ * function could be the below callable struct:
+ * ```
+ * struct ArrayOfArraysSizeFunctor {
+ *   template <size_t OuterArraySize, size_t InnerArraySize, typename T>
+ *   size_t operator()(std::array<std::array<T, InnerArraySize>, OuterArraySize>
+ *                         array_of_arrays) noexcept {
+ *     return OuterArraySize * InnerArraySize;
+ *   }
+ * };
+ * ```
+ * The `size` function must take the single argument of the applicable
+ * container, and should return a `size_t`. This follows the convention of
+ * `std::size()` as of C++17.
+ * \note `std::complex` are regarded as non-indexable (despite a predictable
+ * memory layout), so this function will return 1 for a `std::complex` of a
+ * fundamental type)
  */
-template <class T>
-SPECTRE_ALWAYS_INLINE auto get_size(const T& t) noexcept -> decltype(t.size()) {
-  return t.size();
-}
-
-/// \cond
-template <class T, Requires<std::is_fundamental<T>::value> = nullptr>
-SPECTRE_ALWAYS_INLINE size_t get_size(const T& /*t*/) noexcept {
```

Comment:
For the implementation details in the elementwise approximation, I want a default `Size` and `At` callable anyway, so unless I misunderstand, adding the overload would only add code.

Reply:
I'm saying if you add an overload the default doesn't need to be there because it will fall back to the other case (which is your current default) anyway :)  The default behavior is what currently happens, and if you add an argument that doesn't have a default then the compiler can disambiguate between the two and you get your normal default behavior

Reply:
I understand the process you're suggesting, but if the default `At` and `Size` structs are also used in the element-wise test, as defaults in much the same way a possibly user-supplied `Approx` is taken.
So, without these default structs, I would also need to make duplicate default overloads in the element-wise test utility as well, which I suspect is a net code gain as compared to the current. 

Ultimately, I would like these default `At` and `Size` structs somewhere: either here or in the `TestHelpers`, and it seems preferable to put them here and use them as defaults.

Reply:
After looking through the code more I think the current implementation is more difficult to understand than it needs to be and I'm really not a big fan of having global scope structs with the names `At` and `Size` floating around since the name is not overly descriptive in a general context. If you would like them to be the defaults in the tests I'm fine with sticking them in a `detail` namespace there so you don't need the various permutations of different defaults.

Reply:
I'm fine with that design, though it is notable that doing so does change functionality. The current will always return 1 for a fundamental size and itself for a fundamental at. The suggested change insists that a user-defined function must also handle the fundamental case itself as well. 

I don't mind this much, but it means that the result will be almost identical to the original implementation, with the full at and size function definitions in the `TestHelpers` detail namespace.

Reply:
Fair enough. How about this:
```cpp
/// \ingroup UtilitiesGroup
/// \brief Callable struct which retrieves the `t.size()` for operand `t`. This
/// will cause a compiler error if no such function exists.
struct GetContainerSize {
  template <typename T>
  SPECTRE_ALWAYS_INLINE decltype(auto) operator()(const T& t) noexcept {
    return t.size();
  }
};

/// \ingroup UtilitiesGroup
/// \brief Callable struct for the subscript operator. Returns `t[i]`
struct GetContainerElement {
  template <typename T>
  SPECTRE_ALWAYS_INLINE decltype(auto) operator()(T& t,
                                                  const size_t i) noexcept {
    return t[i];
  }
};

namespace ContainerHelpersDetail {
// implementation struct for get_element and get_size
template <bool IsFundamentalOrComplexOfFundamental>
struct GetImpls;

template <>
struct GetImpls<true> {
  template <typename T, typename SubscriptFunction>
  static SPECTRE_ALWAYS_INLINE decltype(auto) get_element(
      T& t, const size_t /*i*/, const SubscriptFunction /*at*/) noexcept {
    return t;
  }

  template <typename T, typename SizeFunction>
  static SPECTRE_ALWAYS_INLINE size_t
  get_size(const T& /*t*/, const SizeFunction /*size*/) noexcept {
    return 1;
  }
};

template <>
struct GetImpls<false> {
  template <typename T, typename SubscriptFunction>
  static SPECTRE_ALWAYS_INLINE decltype(auto) get_element(
      T& t, const size_t i, SubscriptFunction at) noexcept {
    return at(t, i);
  }

  template <typename T, typename SizeFunction>
  static SPECTRE_ALWAYS_INLINE decltype(auto) get_size(
      const T& t, SizeFunction size) noexcept {
    return size(t);
  }
};
}  // namespace ContainerHelpersDetail

template <typename T, typename SubscriptFunction = GetContainerElement>
SPECTRE_ALWAYS_INLINE decltype(auto) get_element(
    T& t, const size_t i, SubscriptFunction at = At{}) noexcept {
  return ContainerHelpersDetail::GetImpls<(
      tt::is_complex_of_fundamental_v<std::remove_cv_t<T>> or
      cpp17::is_fundamental_v<std::remove_cv_t<T>>)>::get_element(t, i,
                                                                  std::move(
                                                                      at));
}

template <typename T, typename SizeFunction = GetContainerSize>
SPECTRE_ALWAYS_INLINE decltype(auto) get_size(
    const T& t, SizeFunction size = SizeFunction{}) noexcept {
  return ContainerHelpersDetail::GetImpls<(
      tt::is_complex_of_fundamental_v<T> or
      cpp17::is_fundamental_v<T>)>::get_size(t, std::move(size));
}
```

- Fewest class creations possible
- Avoids SFINAE wherever possible
- No overly generic names in the global namespace
- Does not require users writing the structs to handle the fundamental case
- Maximally `const`
- methods are `static` on the `Impl` to avoid constructing them

Reply:
Okay, this version I'm happy with :+1: .